Update README (#11)
heroku/buildpacks: README.md

  <img alt="Heroku Cloud Native Buildpacks Logo" src="./logo.png" width="100%">
</p>

Create a production-ready container image for your application with a [single command](TUTORIAL.md); no Dockerfile required. [Try Heroku Cloud Native Buildpacks today](TUTORIAL.md).

## What is a Cloud Native Buildpack?

> In 2012, Heroku introduced the world to the declarative build system now known as [classic buildpacks](https://blog.heroku.com/buildpacks). Twelve years later, we've joined forces with the Cloud Native Compute Foundation (CNCF) to bring the next evolution in building reliable and consistent images to everyone with [Cloud Native Buildpacks](https://buildpacks.io).

Reviewer: Heroku joined forces with CNCF earlier than 2024 though? Maybe "Twelve years later, we've joined" can be rephrased with either "A few years ago, we joined..." or just "Since then, we've joined"?

Reviewer 2: I like adding the link to the blog here.

The use of “classic” is an internal phrasing still, so needs some attention (just “known as buildpacks”)

Then I tend to agree with Rune that the date and time period don’t add much value.

Author: The goal is to establish a timeline and a pattern. I want to say/show:

We've been here for years, rocking our peers ([not putting suckers in fear](https://genius.com/Ll-cool-j-mama-said-knock-you-out-lyrics)). I.e. We're not following some random standard. We INVENTED the concept. This isn't a weekend hack project, these are the newest branches of an already strong tree. We open-sourced the original, and we're back at it. We are open-sourcing this one.

While we've been a part of CNCF for some time, I'm intentionally choosing to pick a milestone that gives us a larger number. The milestone of these preview releases. And I'm calling out the delta in time as I feel it's emotionally more weighted "twelve years" has an emotional impact, while asking people to do 2024-2012 math in their head does not.

This "call back to history" is a bit of a Heroku-ism and you can see it in a lot of our more grand blog posts. 

If you don't like the specific numbers, some other phrases that might convey similar emotional weight like "Over a decade ago" or "Heroku, founded nearly two decades ago, introduce the world to buildpacks..." (2007, so it's 3 years off, not sure if it's close enough for "nearly"). The goal is to convey: Permanance, weight, commitment to community, and leadership in the space.

> The use of “classic” is an internal phrasing still, so needs some attention (just “known as buildpacks”)

I think for current customers it's confusing to not have a name for both. I'm referring to them as "classic" here https://github.com/heroku/buildpacks-ruby/blob/main/docs/upgrading.md. Other terminology would be "v2" or "heroku specific" or "proprietary standard" which all seem not as well positioned as "classic." For the kubecon crowd, they won't need a differentiator, but I think for Heroku customers calling them "buildpacks" and "Cloud Native Buildpacks" is confusing.

Author: For "classic", we could say "the first" or "the original" or something similar. The usage is more like an adjective than a noun. I'm wanting some small acknowledgment that there's a difference between "<no descriptor> buildpacks" and "CNB buidpacks" (which now strikes me a being like saying "ATM machine").

Reviewer: I'm not against saying that we've been working on (and indeed, invented) the concept of buildpacks for many years at all - or using this to show our commitment in this area.

It's not really about the numbers, I guess I just found it a bit odd to say that _"Twelve years later, we've joined forces with the Cloud Native Compute Foundation (CNCF)"_, when we actually [joined forces with them in 2018](https://blog.heroku.com/buildpacks-go-cloud-native) - if that makes sense?

Btw CNCF is "Cloud Native Computing Foundation" :)

Reviewer: How about this:

```suggestion
Twelve years ago, Heroku introduced the world to the declarative build system known as [buildpacks](https://blog.heroku.com/buildpacks). Since then, we've joined forces with the Cloud Native Compute Foundation (CNCF) to bring the next evolution in building reliable and consistent images to everyone with [Cloud Native Buildpacks](https://buildpacks.io).
```

Cloud Native Buildpacks (CNBs) are an [open specification for building and running container images](https://buildpacks.io/) by the CNCF. Heroku Cloud Native Buildpacks implement the CNB spec, are open-source, can be used independently of the Heroku platform, and are available now as a preview.

### Philosophy

> While a Dockerfile is procedural, CNBs are declarative. Give us your application, and we'll give you an image that can run it. Don't believe us? [Try it out today](TUTORIAL.md).

Reviewer: > While a Dockerfile is procedural, CNBs are declarative

I feel like this calls for a (likely very interesting) in-depth exploration/definition of the philosophical differences between procedural and declarative approaches as applied in this context (and "with the app developer in mind"). Without that, I think there's a likelihood this rephrasing will lead to more confusion than clarity for someone just starting to learn about CNB.

Reviewer 2: I tend to agree I think this section is in pretty good shape as is.

Reviewer 2: ok. Thanks for the context!

Keeping in mind this is the "philosophy" section, so I'm consuming this as an ideal not a guarantee. Using terms like "easy" and "best-practices" are ok imo b/c it's a philosophy. And yes, people will judge if it's those things and then feed that back so we can improve towards our goal.

That being said your points are valid. Maybe just combine the best parts of each?

> Heroku CNBs are written with the app developer in mind. We want it to be as easy and simple as possible to transform an app into an OCI image. While a Dockerfile is procedural, CNBs are declarative. Give us your application, and we'll give you an image that can run it. Don't believe us? [Try it out today](TUTORIAL.md).

Reviewer: I knew there was a very interesting explanation (and lots of thoughts) behind this change! ;-)

I also agree that "highly opinionated" and "industry best-practices" are used too much/without substantiating what that means.

I like [@jonnymacs's suggestion](https://github.com/heroku/buildpacks/pull/11#discussion_r1530844638) to combine them -- the first two lines explain our goal/mission, and _"While a Dockerfile is procedural, CNBs are declarative"_ is how we make that happen. The following three sentences underscore both the declarative approach, and encourages/challenges people to see if we've achieved our goals :)

Seems like a reasonable compromise?

Author: I took Jonny's change wholesale.

We believe:
- Building OCI images shouldn't be a pain.
- Developers should spend time writing application code, not endless container configurations.

> -Community standards and build experiences that **just work** like you expect them to.

Reviewer: ```suggestion
- Community standards and build experiences that **just work** like you expect them to.
```

## See it in action

> This section will use Heroku's Buildpacks to transform a Node.js app into a container image. The process for other languages is identical; you are encouraged to try it with your application!

Reviewer 2: 👍

### Prerequisites

#### Docker

If you don't have it already, you need to install Docker. Refer to the Docker documentation on how to install it for your operating system: https://docs.docker.com/engine/install/

#### pack

> To run our buildpack locally, we will use `pack`, a tool maintained by the Cloud Native Buildpacks project. Follow their documentation to [install the pack CLI](https://buildpacks.io/docs/install-pack/).

Reviewer 2: 👍

### Create the app image

> [!NOTE]

> > For a full hands-on getting started experience [follow our getting started tutorial](TUTORIAL.md).

Reviewer 2: 👍

First, we need an application to produce an image. We will use [Heroku's getting started app for Node.js](https://github.com/heroku/node-js-getting-started), but any other app that uses a supported ecosystem will work. Clone the application:

```
$ git clone https://github.com/heroku/node-js-getting-started.git
$ cd node-js-getting-started
```

In your application directory, run the `pack build` command. No additional steps are needed! The required buildpacks for your app will be detected automatically:

```

> $ pack build my-app-image --builder heroku/builder:22 --path .

Reviewer: Might as well drop `--path .` if we've already changed directory to `node-js-getting-started`

```suggestion
$ pack build my-app-image --builder heroku/builder:22
```

Author: From experience, people will run this in the wrong directory. Either they won't CD, or maybe they cloned it into a different directory, or a thousand other problems. The way to correct for this in a tutorial is to give a "check" step (as I did in the tutorial by showing the contents of `ls`) but this is a minimal example. So if this command fails  I want the command to act as a bit of "oh, wait...I'm passing in arguments, is the path correct?" marker.

As a much weaker argument, I've worked with tools that cannot handle a path to a different directory; you can ONLY use them in the current directory. I feel the three arguments here—image name, builder name, and path—are likely the three most common, and I would like to demonstrate them.

As a side note: I would actually like to encourage people to set heroku as the default builder as I do in the tutorial. However, I like that it's explicit here. We would do well to add a "Note" to this section asking people to set their default builder, and also updating the tutorial to show people how they can use `--path`.

Reviewer: Fine by me, really - though I'd argue that if you wanted to show people how they can use `--path` in a way that isn't unnecessary, the existing text already did just that (there's no need to `cd` into the directory, as we don't do anything else with the app in this short example):

```
git clone https://github.com/heroku/node-js-getting-started.git

pack build my-app-image --builder heroku/builder:22 --path node-js-getting-started
```

Reviewer: I still think the `git clone` command can be presented better though (I moved it to a code block/it's own line here https://github.com/heroku/buildpacks/pull/13/commits/aca76c051f7fb5c1f4775bd1c83fe46771560762, but closed that PR to not conflict with your changes here :))

Author: For this case I like the `git clone` without the `cd` and keeping the `--path` (effectively what it is on main but `git clone` is on it's own line.

```

Once the image is built, you can run it with the tools you're already comfortable with, like the `docker` CLI, for a demonstration of how [read out getting started tutorial](TUTORIAL.md).

## Discussions

> We want to provide great user experiences and can only do that with your help. Please try our Cloud Native Buildpacks and let us know how it went. What went well? What could be better?  Let us know on GitHub in the [discussions tab of this repository](https://github.com/heroku/buildpacks/discussions). Feel free to open a new discussion if you have any questions, feedback, or other input - we want to hear from you!

Reviewer 2: 👍

Reviewer: When I read the PR that introduced "Overarching", I had to look it up to verify it was a valid English word :) Turns out it is, but this is a better way to phrase the sentiment.

Think we can probably drop the last sentence/"call to action" now though (seeing as we've already solicited feedback/discussions)? 

```suggestion
We want to provide great user experiences and can only do that with your help. Please try our Cloud Native Buildpacks and let us know how it went. What went well? What could be better?  Let us know on GitHub in the [discussions tab of this repository](https://github.com/heroku/buildpacks/discussions)!
```

### libcnb.rs

> Heroku buildpacks are implemented in Rust for speed, safety, and security. We developed and open-sourced a framework for writing Cloud Native Buildpacks in Rust called [libcnb.rs](https://github.com/heroku/libcnb.rs).

Reviewer 2: 👍

You can use [libcnb.rs](https://github.com/heroku/libcnb.rs) to implement your own Cloud Native Buildpacks.

### Builder Images

> A CNB builder image is a packaged set of buildpacks, [base images](https://github.com/heroku/base-images), and a lifecycle binary for orchestrating the build process. Find [Heroku's CNB Builder Images here](https://github.com/heroku/cnb-builder-images).

Reviewer 2: 👍

### Procfile Buildpack

> A `Procfile` is a [Heroku standard for declaring process types on your application](https://devcenter.heroku.com/articles/procfile). Our Procfile Cloud Native Buildpack brings this great Heroku experience to the Cloud Native world, and the source is available [here on GitHub as well](https://github.com/heroku/buildpacks-procfile).

Reviewer 2: 👍
